drivers/soundwire/intel.c

```diff
@@ -360,10 +360,23 @@ static int intel_link_power_up(struct sdw_intel *sdw)
 	 * tables. The values reported are based on either 24MHz
 	 * (CNL/CML) or 38.4 MHz (ICL/TGL+).
 	 */
-	if (prop->mclk_freq % 6000000)
-		syncprd = SDW_SHIM_SYNC_SYNCPRD_VAL_38_4;
-	else
+	if (prop->mclk_freq % 6000000) {
+		if (prop->mclk_freq % 3072000)
+			syncprd = SDW_SHIM_SYNC_SYNCPRD_VAL_38_4;
+		else
+			syncprd = SDW_SHIM_SYNC_SYNCPRD_VAL_24_576;
+	} else {
 		syncprd = SDW_SHIM_SYNC_SYNCPRD_VAL_24;
+	}
+
+	if (sdw->link_res->cardinal_clock_capable) {
+		u32 mlcs_mask;
+
+		link_control = intel_readl(shim, SDW_SHIM_LCTL);
+		mlcs_mask = FIELD_PREP(SDW_SHIM_LCTL_MLCS_MASK, SDW_SHIM_LCTL_MLCS_CARDINAL_CLK);
+		link_control |= mlcs_mask;
```

Comment:
@plbossart Should we clear SDW_SHIM_LCTL_MLCS_MASK first? For example, if link_control bit 29 to 27 is 0x011 (MCLK input), and mlcs_mask = 0x001, then link_control bit 29 to 27 will be still 0x011 after link_control |= mlcs_mask, right?

Reply:
done, I used u32_replace_bits and that does the trick.